### PR TITLE
goal: Do not add lower versions to upgrade jobs

### PR DESCRIPTION
**Pull request by m-blaha**
For UPGRADE* actions, do not add packages that are older than the
currently installed versions to the libsolv job. This decreases the
number of potential solutions and helps resolve some edge cases.

See https://issues.redhat.com/browse/RHEL-1448 for one of those edge
cases.

Test: https://github.com/rpm-software-management/ci-dnf-stack/pull/1380

### CODE REVIEW

File: libdnf5/base/goal.cpp [hunk 1]
```diff
@@ -509,6 +509,13 @@ GoalProblem Goal::Impl::add_specs_to_goal(base::Transaction & transaction) {
                         base, query, settings.advisory_filter.value(), cfg_main.get_obsoletes_option().get_value());
                 }
 
+                // Remove packages of versions lower than the installed ones.
+                // Substract downgrades instead of filter_upgrades() to keep
+                // installed packages in the query.
+                rpm::PackageQuery tmp_downgrades(query);
```
Comment (j-mracek) [inline libdnf5/base/goal.cpp:515]: I am quite curious, why you want to keep packages with the same version as installed in the query?

Comment (m-blaha) [reply libdnf5/base/goal.cpp:515]: Well, I don't :) I've just overlooked it. Thanks for noticing.

Comment (m-blaha) [reply libdnf5/base/goal.cpp]: Fixed.

Comment (m-blaha) [reply libdnf5/base/goal.cpp]: > I recall that "dnf upgrade" reinstalls packages which were originally installed unsigned from a local build and then they appeared properly signed in a repository. That could warrant leaving same versions in the query.

I need to check this. We do not have a test for it, I suppose?

Comment (m-blaha) [reply libdnf5/base/goal.cpp]: @ppisar it seems that you are refering to behaviour described in this libsolv issue: https://github.com/openSUSE/libsolv/issues/287
The trigger for re-installation was not whether the package was or was not signed, but more generally that the package was changed (for example has different build time).
This behavior was changed in dnf4 here: https://github.com/rpm-software-management/libdnf/pull/786 (and the same behavior we have in dnf5).
So dnf4 does not reinstall package with the same nevra but different buildtime for quite a long time. But distrosync command should trigger the reinstallation.

File: libdnf5/base/goal.cpp [hunk 2]
```diff
@@ -71,6 +71,19 @@ void add_obsoletes_to_data(const libdnf5::rpm::PackageQuery & base_query, libdnf
     data |= obsoletes_query;
 }
 
+/// Removes from the query available packages with versions lower than or equal
+/// to the installed ones. Similar to filter_upgrades(), but retains also
+/// installed packages and packages whose names are not installed.
+static void remove_older_versions(libdnf5::rpm::PackageQuery & query) {
+    libdnf5::rpm::PackageQuery installed(query.get_base());
```
Comment (j-mracek) [inline libdnf5/base/goal.cpp:78]: I would suggest to initialize query without excluded packages

`libdnf5::rpm::PackageQuery installed(query.get_base(), rpm::PackageQuery::ExcludeFlags::IGNORE_EXCLUDES)`

File: libdnf5/base/goal.cpp [hunk 3]
```diff
@@ -71,6 +71,19 @@ void add_obsoletes_to_data(const libdnf5::rpm::PackageQuery & base_query, libdnf
...
+    installed.filter_installed();
```
Comment (j-mracek) [inline libdnf5/base/goal.cpp:79]: This is very weak suggestion because it does not change the result.
What about to add additional filter  - `latest()`? Installonly packages will be handled in the next query only once.

File: libdnf5/base/goal.cpp [hunk 4]
```diff
@@ -509,6 +522,8 @@ GoalProblem Goal::Impl::add_specs_to_goal(base::Transaction & transaction) {
                         base, query, settings.advisory_filter.value(), cfg_main.get_obsoletes_option().get_value());
                 }
 
+                remove_older_versions(query);
```
Comment (j-mracek) [inline libdnf5/base/goal.cpp:525]: May I ask you to reference the bug or write a note why it is important?

File: libdnf5/base/goal.cpp (hunk 2 above)
Comment (m-blaha) [reply libdnf5/base/goal.cpp]: Makes sense, fixed.

File: libdnf5/base/goal.cpp (hunk 3 above)
Comment (m-blaha) [reply libdnf5/base/goal.cpp:83]: Added. (although I did not measure impact on performance here).

File: libdnf5/base/goal.cpp (hunk 4 above)
Comment (m-blaha) [reply libdnf5/base/goal.cpp:530]: Sure. Added to `remove_older_versions()` function description.

File: libdnf5/base/goal.cpp (hunk 1 above)
Comment (ppisar) [reply libdnf5/base/goal.cpp]: Yes, that libsolv ticket describes my experience. If DNF stopped doing it, the then j-mracek's proposal is fine.

**Review by j-mracek — APPROVED** [review]

**Comment by m-blaha** [comment]
Looks like some tests with repo priorities started failing. I'll investigate.

**Comment by j-mracek** [comment]
We have a problem, if we do this we will have a problem with repo priority test cases. Basically we have two options, we can keep all packages from proiority repo and then apply removal of old packages (logic from check update)

**Comment by m-blaha** [comment]
> We have a problem, if we do this we will have a problem with repo priority test cases. Basically we have two options, we can keep all packages from proiority repo and then apply removal of old packages (logic from check update)

I've added code to pre-select upgrade candidates considering also repository priorities. However, I'm starting to feel that we might be exceeding our responsibilities by intervening in matters that should be handled by libsolv. To be honest, I'm leaning towards closing this PR and simply documenting in the upgrade code why including all versions of the package in the libsolv job is important.

**Comment by m-blaha** [comment]
@ppisar unfortunately, the `--best` does not help here and the result remains the same.
Unversioned obsoletes are obstacle, that's hard to get around. I don't see other solution than rename the package and start again from scratch.
